---
title: "Notebook to reproduce Figure 3d & S6b"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
author: "Yiyan Wu"
params:
  output_dir: "output" 
  archr_proj: "data/multiome_results/10_ArchR/CellClass_L5_2"
  archr_threads: 8
  genome_version: "hg38"
  TFs_of_interest: "data/Neuronal_TF.csv"
  TF_deviations: "data/TF_deviations.rds"
  TF_exp_mtx: "data/TF_exp_mtx.csv"
output: html_document
---


```{r, rmarkdown-setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here(), warning = FALSE)
```

```{r, temp}
setwd("/cluster/projects/gaitigroup/Users/Yiyan/GBM_publication/gbm_manuscript")
params <- list(
    output_dir = "output",
    archr_proj = "data/multiome_results/10_ArchR/CellClass_L5_2",
    archr_threads = 8,
    genome_version = "hg38",
    celltype_column = "CellClass_L5_2",
    patient_column = "Patient",
    region_column = "Region",
    confidence_column = "Confident_Annotation",
    include_types = "Malignant_NPC1,Malignant_NPC2,Malignant_OPC,Invasive-high_OPC_NPC1,Malignant_MES_HYP,Malignant_MES_INT,Malignant_MES_AST,Malignant_AC",
    exclude_types = NULL,
    include_regions = NULL,
    TFs_of_interest = "data/Neuronal_TF.csv",
    TF_deviations = "data/TF_deviations.rds",
    TF_exp_mtx = "data/TF_exp_mtx.csv"
)

```

## Prepare data
```{r, prep-data}
pacman::p_load(GaitiLabUtils, GBMutils, glue, data.table, tidyverse, stringr, ArchR, BSgenome.Hsapiens.UCSC.hg38, cowplot, dplyr, data.table, reshape, ggpubr, rstatix, tibble, ggrepel, ggplot2, circlize, ComplexHeatmap, viridis, Seurat)

# Load ArchR project
addArchRThreads(threads = as.numeric(params$archr_threads))
addArchRGenome(genome = params$genome_version)
archr_proj <- loadArchRProject(params$archr_proj)

# Filter out the cell types
cell_type_column <- paste0("Seurat_", params$celltype_column)
patient_column <- paste0("Seurat_", params$patient_column)
region_column <- paste0("Seurat_", params$region_column)
confidence_column <- paste0("Seurat_", params$confidence_column)

if(!is.null(params$include_types)){
  include_types <- str_split(params$include_types, ",")[[1]]
  archr_proj <- archr_proj[which(getCellColData(archr_proj, select = cell_type_column, drop = TRUE) %in% include_types)]
}else if(!is.null(params$exclude_types)){
  exclude_types <- str_split(params$exclude_types, ",")[[1]]
  archr_proj <- archr_proj[which(!(getCellColData(archr_proj, select = cell_type_column, drop = TRUE) %in% exclude_types))]
}

archr_proj <- archr_proj[which(getCellColData(archr_proj, select = confidence_column, drop = TRUE) == "TRUE")]

if (!is.null(params$include_regions)) {
  include_regions <- str_split(params$include_regions, ",")[[1]]
  archr_proj <- archr_proj[which(getCellColData(archr_proj, select = region_column, drop = TRUE) %in% include_regions)]
}

# TF with high deviation from ChromVAR in inv high and high correlation to gene expression
# Identifying deviant TF motifs
seGroupMotif <- getGroupSE(ArchRProj = archr_proj, useMatrix = "MotifMatrix", groupBy = "Seurat_CCI_CellClass_L2_2")
seZ <- seGroupMotif[rowData(seGroupMotif)$seqnames == "z", ]

# Identify the maximum delta in z-score between Invasive-high OPC/NPC1 and Progenitor_like
rowData(seZ)$Inv_delta <- assay(seZ)[, "Invasive-high OPC/NPC1"] - assay(seZ)[, "Progenitor_like"]

# Identifying correlated TF motifs and TF expression
corGEM_MM <- correlateMatrices(
  ArchRProj = archr_proj,
  useMatrix1 = "GeneExpressionMatrix",
  useMatrix2 = "MotifMatrix",
  reducedDims = "LSI_Combined"
)

# Add maximum delta deviation to the correlation data frame
corGEM_MM$Inv_delta <- as.numeric(rowData(seZ)[match(corGEM_MM$MotifMatrix_name, rowData(seZ)$name), "Inv_delta"])

# Get the motifs to plot
motifs_df <- fread(params$TFs_of_interest, header = TRUE)
motifs_to_plot <- unique(motifs_df$TF[motifs_df$Group == "ScenicPlus"])

# Filter out duplicated TFs
corGEM_MM <- corGEM_MM[order(abs(corGEM_MM$cor), decreasing = TRUE), ]
corGEM_MM <- corGEM_MM[which(!duplicated(gsub("\\-.*", "", corGEM_MM$MotifMatrix_name))), ]

# TF to label in the plot
high_SCENIC_TF <- motifs_df$TF[motifs_df$Subgroup == "Invasive-high OPC/NPC1"]
high_SCENIC_TF <- paste0(high_SCENIC_TF, "_")

low_SCENIC_TF <- motifs_df$TF[motifs_df$Subgroup == "Differentiated_like"]
low_SCENIC_TF <- paste0(low_SCENIC_TF, "_")
low_SCENIC_TF <- sort(low_SCENIC_TF)

# NPC/OPC markers
NPC_OPC_markers <- c("SOX2", "SOX4", "SOX11", "OLIG1", "OLIG2")
NPC_OPC_markers <- paste0(NPC_OPC_markers, "_")

# AP-1 family TFs
AP1_TF <- c("FOS", "FOSB", "FOSL1", "FOSL2", "FOSL2", "JUN", "JUNB", "JUND")

# Motifs to highlight in ChromVAR
Positive_TF <- corGEM_MM$MotifMatrix_name[corGEM_MM$cor > 0 & corGEM_MM$Inv_delta > quantile(corGEM_MM$Inv_delta, 0.9)]
Positive_TF <- Positive_TF[!is.na(Positive_TF)] # remove NA
Positive_TF_name <- substr(Positive_TF, 1, regexpr("_", Positive_TF) - 1) # remove number after _
Positive_TF_name <- paste0(Positive_TF_name, "_")

# Find high_SCENIC_TF intersect with Positive_TF
motifs_to_plot <- high_SCENIC_TF[high_SCENIC_TF %in% Positive_TF_name]

# Find TFs in each category
high_SCENIC_TF_to_plot <- high_SCENIC_TF[!sapply(high_SCENIC_TF, function(x) any(sapply(motifs_to_plot, function(y) grepl(y, x))))] # Identified in SCENIC but not in ChromVAR
high_SCENIC_TF_to_plot <- corGEM_MM$MotifMatrix_name[sapply(corGEM_MM$MotifMatrix_name, function(x) any(sapply(high_SCENIC_TF_to_plot, function(y) grepl(y, x))))]
motifs_to_plot <- Positive_TF[sapply(Positive_TF, function(x) any(sapply(motifs_to_plot, function(y) grepl(y, x))))] # Identified in both SCENIC and ChromVAR

# Label the TFs
corGEM_MM$TFRegulator <- "Not significant"
corGEM_MM$TFRegulator[corGEM_MM$MotifMatrix_name %in% high_SCENIC_TF_to_plot] <- "Candidate identified in SCENIC+"
corGEM_MM$TFRegulator[corGEM_MM$MotifMatrix_name %in% motifs_to_plot] <- "Putative regulator"
corGEM_MM$TFRegulator[sapply(corGEM_MM$MotifMatrix_name, function(x) any(sapply(NPC_OPC_markers, function(y) grepl(y, x))))] <- "NPC/OPC markers"
corGEM_MM$TFRegulator[sapply(corGEM_MM$MotifMatrix_name, function(x) any(sapply(AP1_TF, function(y) grepl(y, x))))] <- "AP-1 family TFs"

# Fix MotifMatrix_name - delete everything after MotifMatrix_name
corGEM_MM$MotifMatrix_name <- substr(corGEM_MM$MotifMatrix_name, 1, regexpr("_", corGEM_MM$MotifMatrix_name) - 1)
```

## Visualization
```{r, vis-1, fig.width = 7, fig.height = 7.5}
# Visualize the correlation between TF motif and gene expression
ggplot(data.frame(corGEM_MM), aes(cor, Inv_delta, color = TFRegulator)) +
  geom_point(data = data.frame(corGEM_MM[corGEM_MM$TFRegulator == "Not significant", ]), aes(cor, Inv_delta, color = TFRegulator), size = 1, alpha = 0.75) +
  geom_point(data = data.frame(corGEM_MM[corGEM_MM$TFRegulator != "Not significant", ]), aes(cor, Inv_delta, color = TFRegulator), size = 2) +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  geom_hline(yintercept = quantile(corGEM_MM$Inv_delta, 0.9), lty = "dashed", color = "darkgrey") +
  scale_color_manual(values = c(
    "Not significant" = "darkgrey",
    "Putative regulator" = "#2B7095",
    "Candidate identified in SCENIC+" = "#EDAE49FF",
    "NPC/OPC markers" = "#7C9EB5",
    "AP-1 family TFs" = "#C05E00"
  )) +
  geom_label_repel(data = data.frame(corGEM_MM[corGEM_MM$TFRegulator != "Not significant", ]), aes(label = MotifMatrix_name), size = 3, nudge_x = 0.1, nudge_y = 0.1, max.overlaps = 10) +
  labs(
    y = "TF motif accessibility difference between \ninvasive-high OPC/NPC1 and progenitor-like (Î”z-score)",
    x = "Correlation of TF motif accessibility and TF expression"
  ) +
  scale_y_continuous(
    expand = c(0, 0),
    limits = c(min(corGEM_MM$Inv_delta) * 1.05, max(corGEM_MM$Inv_delta) * 1.05)
  ) +
  scale_x_continuous(
    expand = c(0, 0),
    limits = c(-1, 1)
  ) +
  GBM_theme() +
  theme(
    legend.position = "bottom",
    legend.title = element_blank()
  )

ggsave(paste0(params$output_dir, "/corGEM_MM_TF_Regulator.pdf"), width = 7, height = 7.5)
```


```{r, vis-2, fig.height = 80, fig.width = 5}
# ---------------------------- Scatter plot for each TF ---------------------------- #
p <- readRDS(params$TF_deviations)

motifs_to_plot <- c("TCF12", "TCF4", "TCF3", "ASCL1", "ZEB1", "KLF12", "FOSL1", "FOSL2")

df_neuronal <- NULL
for (i in 1:length(p)) {
  p1 <- p[[i]]

  # get the name of the TF
  curr_TF <- names(p)[i]
  curr_TF <- strsplit(curr_TF, "_")[[1]][1]
  curr_TF <- strsplit(curr_TF, "z:")[[1]][2]
  print(curr_TF)

  if (!(curr_TF %in% motifs_to_plot)) next

  # add the data to the dataframe
  df_neuronal <- rbind(df_neuronal, p1$data %>% mutate(TF = curr_TF))
}

# Load the expression data for the TFs
exp_mat <- fread(params$TF_exp_mtx)
exp_mat$archr_cellnames <- sub("-1.*", "", exp_mat$archr_cellnames)

for (TF_name in unique(df_neuronal$TF)) {
  print(paste0("Currently analyzing: ", TF_name))
  curr_TF_chromVAR <- df_neuronal[df_neuronal$TF == TF_name, ] %>%
    rownames_to_column(var = "archr_cellnames")
  curr_TF_chromVAR$archr_cellnames <- sub("-1.*", "", curr_TF_chromVAR$archr_cellnames)
  curr_TF_exp <- exp_mat %>%
    filter(archr_cellnames %in% curr_TF_chromVAR$archr_cellnames) %>%
    select(archr_cellnames, TF_name)
  curr_df <- merge(curr_TF_chromVAR, curr_TF_exp, by = "archr_cellnames") %>%
    mutate(
      Sample = sub("#.*", "", archr_cellnames),
      Barcode = sub(".*#", "", archr_cellnames)
    )

  # Aggregate at cell type level
  curr_df <- curr_df %>%
    dplyr::group_by(x) %>%
    dplyr::summarise(mean_cell_type_gex = mean(get(TF_name)),
              mean_cell_type_acc = mean(y),
              total_n = n())

  # Rename cell type
  curr_df <- curr_df %>%
    mutate(x = case_when(
      x == "Malignant_NPC1" ~ "NPC1-like",
      x == "Malignant_NPC2" ~ "NPC2-like",
      x == "Malignant_OPC" ~ "OPC-like",
      x == "Invasive-high_OPC_NPC1" ~ "Invasive-high OPC/NPC1",
      x == "Malignant_AC" ~ "AC-like",
      x == "Malignant_MES_AST" ~ "MES-AC-like",
      x == "Malignant_MES_INT" ~ "MES-INT-like",
      x == "Malignant_MES_HYP" ~ "MES-HYP-like"
    ))

  color_palette <- c(
    "NPC1-like" = "#86CCE8",
    "NPC2-like" = "#ADD7E5",
    "OPC-like" = "#AECFA5",
    "AC-like" = "#DFB63B",
    "MES-AC-like" = "#F37F72",
    "MES-INT-like" = "#B02425",
    "MES-HYP-like" = "#F1634B",
    "Invasive-high OPC/NPC1" = "#2B7095"
  )

  # Plot with legend
  ggplot(curr_df, aes(x = mean_cell_type_gex, y = mean_cell_type_acc)) +
    geom_point(aes(size = total_n, color = "#000000", fill = x), shape = 21, stroke = 1) + # Set stroke and shape here
    geom_text_repel(aes(label = x), size = 5) +
    scale_color_manual(values = color_palette) + # Use custom color palette
    scale_fill_manual(values = color_palette) + # Use custom color palette for fill
    labs(x = "Scaled mean gene expression", y = "Motif accessibility (mean z-score)") +
    GBM_theme() +
    theme(
      legend.position = "bottom",
      legend.title = element_blank()
    ) +
    ggtitle(TF_name)
  ggsave(paste0(params$output_dir, "/Neuronal_motif_deviation_score_scatterplot_legend_", TF_name, ".pdf"), width = 5, height = 5.5)
}
```


## Session Info
```{r, session-info, echo = FALSE}
sessionInfo()
```
